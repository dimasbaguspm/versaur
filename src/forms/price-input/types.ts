import type { InputHTMLAttributes } from 'react'
import type { TextInputProps } from '@/forms/text-input/types'

/**
 * Props for PriceInput component
 * @property value - The price value as a string (raw, unformatted)
 * @property onChange - Called when the value changes (returns raw string)
 * @property allowNegative - Optional: allow negative values (default: false)
 * @property inputProps - Optional: custom input attributes
 */
export interface PriceInputProps
  extends Omit<TextInputProps, 'type' | 'value' | 'onChange'> {
  value: string
  onChange: (value: string) => void
  allowNegative?: boolean
  inputProps?: InputHTMLAttributes<HTMLInputElement>
}
