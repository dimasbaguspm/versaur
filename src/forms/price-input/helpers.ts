/**
 * Helpers for PriceInput component - Rupiah currency formatting and validation
 */

/**
 * Validates if the input string follows valid Rupiah currency format
 * Valid formats:
 * - "10000" (simple digits)
 * - "10.000" (thousands separator)
 * - "10.000,00" (thousands separator with decimal)
 * - "10.000.000,59" (multiple thousands separators with decimal)
 * - "0,59" (zero with decimal)
 * - "10.000," (trailing comma while typing)
 *
 * Invalid formats:
 * - "10...1" (multiple consecutive dots)
 * - "10,1,1" (multiple commas)
 * - "10,,1" (consecutive commas)
 * - "10.000.00" (decimal should use comma, not dot)
 * - ",59" (cannot start with comma)
 * - ".59" (cannot start with dot)
 * - "," (cannot be just comma)
 * - "." (cannot be just dot)
 */
export function isValidRupiahFormat(
  input: string,
  allowNegative: boolean = false
): boolean {
  if (input === '') return true
  if (input === '-' && allowNegative) return true

  // Remove leading minus for negative values
  const cleanInput =
    allowNegative && input.startsWith('-') ? input.slice(1) : input
  if (allowNegative && input.startsWith('-') && cleanInput === '') return true

  // Don't allow starting with separators
  if (cleanInput.startsWith('.') || cleanInput.startsWith(',')) return false

  // Check basic format: digits, dots for thousands, comma for decimal
  if (!/^[\d.,]+$/.test(cleanInput)) return false

  // Split by comma to separate integer and decimal parts
  const parts = cleanInput.split(',')

  // Should have at most 1 comma (decimal separator)
  if (parts.length > 2) return false

  const integerPart = parts[0]
  const decimalPart = parts[1]

  // Validate integer part - don't allow empty (prevents ",59" format)
  if (integerPart === '' || !isValidRupiahIntegerPart(integerPart)) return false

  // Validate decimal part (if exists) - allow empty for intermediate typing like "1000,"
  if (
    decimalPart !== undefined &&
    decimalPart !== '' &&
    !isValidRupiahDecimalPart(decimalPart)
  )
    return false

  return true
}

/**
 * Validates the integer part of Rupiah currency
 * Should follow pattern: digits with dots as thousands separators OR pure digits (not empty)
 * Valid: "0", "10", "10000", "10.000", "1.000.000"
 * Also allows intermediate states: "10.", "10.0", "1.000.", "1.000.0" (while typing)
 * Invalid: "", ".000", "10..000", "10.0000" (more than 3 digits in segment)
 */
function isValidRupiahIntegerPart(part: string): boolean {
  if (part === '') return false // Don't allow empty integer parts

  // If no dots, just validate it's all digits (any length)
  if (!part.includes('.')) {
    return /^\d+$/.test(part)
  }

  // Split by dots (thousands separators)
  const segments = part.split('.')

  // First segment can be 1-3 digits
  if (segments[0] === '' || !/^\d{1,3}$/.test(segments[0])) return false

  // Remaining segments must be exactly 3 digits (except for the last segment which can be incomplete while typing)
  for (let i = 1; i < segments.length; i++) {
    const segment = segments[i]

    // Last segment can be incomplete (0-3 digits) while typing - allow empty for intermediate states like "1.000."
    if (i === segments.length - 1) {
      if (segment !== '' && !/^\d{1,3}$/.test(segment)) return false
    } else {
      // Middle segments must be exactly 3 digits
      if (!/^\d{3}$/.test(segment)) return false
    }
  }

  return true
}

/**
 * Validates the decimal part of Rupiah currency
 * Should be exactly 1 or 2 digits, or empty during intermediate typing
 * Valid: "0", "00", "5", "59", "" (while typing)
 * Invalid: "000", "5a"
 */
function isValidRupiahDecimalPart(part: string): boolean {
  return /^\d{1,2}$/.test(part)
}

/**
 * Sanitizes input by removing invalid characters while preserving valid format
 * This function is more permissive during typing to provide better UX
 */
export function sanitizeRupiahInput(
  input: string,
  allowNegative: boolean = false
): string {
  if (input === '') return ''

  // Handle negative values
  const isNegative = input.startsWith('-')
  if (isNegative && !allowNegative)
    return sanitizeRupiahInput(input.slice(1), false)
  if (isNegative && input.length === 1) return '-'

  const cleanInput = isNegative ? input.slice(1) : input

  // Remove all non-digit, non-dot, non-comma characters
  let sanitized = cleanInput.replace(/[^\d.,]/g, '')

  // Handle multiple commas - keep only the last one
  const commaIndex = sanitized.lastIndexOf(',')
  if (commaIndex !== -1) {
    const beforeComma = sanitized.slice(0, commaIndex).replace(/,/g, '')
    const afterComma = sanitized.slice(commaIndex + 1).replace(/,/g, '')
    sanitized = beforeComma + ',' + afterComma
  }

  // Handle consecutive dots - remove extra dots
  sanitized = sanitized.replace(/\.{2,}/g, '.')

  return (isNegative ? '-' : '') + sanitized
}

/**
 * Converts various Rupiah input formats to a standardized numeric value
 * "10.000,50" -> 10000.50
 * "10000" -> 10000
 * "10.000" -> 10000
 */
export function parseRupiahToNumber(input: string): number {
  if (input === '' || input === '-') return 0

  const isNegative = input.startsWith('-')
  const cleanInput = isNegative ? input.slice(1) : input

  // Replace dots with nothing (remove thousands separators)
  // Replace comma with dot (decimal separator)
  const standardized = cleanInput.replace(/\./g, '').replace(',', '.')

  const parsed = parseFloat(standardized) || 0
  return isNegative ? -parsed : parsed
}

/**
 * Formats a number to Rupiah display format
 * 10000.50 -> "10.000,50"
 * 10000 -> "10.000"
 */
export function formatNumberToRupiah(value: number): string {
  if (value === 0) return ''

  const isNegative = value < 0
  const absValue = Math.abs(value)

  const [integerPart, decimalPart] = absValue.toString().split('.')

  // Add thousands separators to integer part
  const formattedInteger = integerPart.replace(/\B(?=(\d{3})+(?!\d))/g, '.')

  // Combine with decimal part if it exists
  const formatted = decimalPart
    ? `${formattedInteger},${decimalPart}`
    : formattedInteger

  return (isNegative ? '-' : '') + formatted
}
