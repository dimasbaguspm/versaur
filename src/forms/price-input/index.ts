export { PriceInput } from './price-input'
export type { PriceInputProps } from './types'
