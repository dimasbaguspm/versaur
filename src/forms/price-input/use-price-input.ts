import { useEffect } from 'react'

/**
 * usePriceInputSync synchronizes the local input value with the parent value
 * @param value - The parent value (string or number)
 * @param setRawValue - Setter for local raw value
 */
export function usePriceInputSync(
  value: string | number,
  setRawValue: (val: string) => void
) {
  useEffect(() => {
    // Only update rawValue if parent value changes (not on every keystroke)
    // If parent value is a number, format it for display
    if (typeof value === 'number' && !isNaN(value)) {
      setRawValue(`${value}`)
    } else if (value === '' || value === null || value === undefined) {
      setRawValue('')
    } else if (typeof value === 'string') {
      // If parent value is a string, use as is
      setRawValue(value)
    }
  }, [value, setRawValue])
}
