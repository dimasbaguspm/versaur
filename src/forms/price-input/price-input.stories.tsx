/**
 * PriceInput stories for Storybook
 * Shows IDR price input with formatting and standard input attributes
 */
import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import { PriceInput } from './price-input'

const meta: Meta<typeof PriceInput> = {
  title: 'Forms/PriceInput',
  component: PriceInput,
}
export default meta

type Story = StoryObj<typeof PriceInput>

/**
 * Default PriceInput usage
 * Demonstrates Rupiah currency format with proper thousands separator (dot) and decimal separator (comma)
 * Valid formats: 10000, 10.000, 10.000,50, 1.000.000,99
 */
export const Default: Story = {
  render: function DefaultStory(args) {
    const [value, setValue] = useState('')
    return (
      <PriceInput
        value={value}
        onChange={v => {
          setValue(v)
          args?.onChange?.(v)
        }}
        label='Amount'
        helperText='Enter price in IDR format (e.g., 10.000 or 10.000,50)'
      />
    )
  },
}

/**
 * PriceInput with error state
 * Shows error message and raw string value usage
 */
export const WithError: Story = {
  render: function WithErrorStory() {
    const [value, setValue] = useState('')
    return (
      <PriceInput
        value={value}
        onChange={setValue}
        label='Amount'
        error='Invalid price'
      />
    )
  },
}
