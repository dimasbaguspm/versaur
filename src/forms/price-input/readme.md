# PriceInput

Versaur PriceInput is a specialized input component for Indonesian Rupiah (IDR) currency entry. It
enforces proper Rupiah formatting conventions with thousands separators (dots) and decimal
separators (commas), while providing real-time validation to prevent invalid input patterns.

## Features

- **Rupiah Format Validation**: Enforces Indonesian currency format (e.g., 10.000 or 10.000,50)
- **Real-time Input Sanitization**: Removes invalid characters and patterns while typing
- **Flexible Input Support**: Accepts various valid formats (10000, 10.000, 10.000,50)
- **Controlled Component**: Raw string value passed to `onChange` for external processing
- **Accessibility**: Full keyboard support with proper input patterns
- **Mobile Optimized**: Numeric input mode for mobile keyboards

## Valid Formats

- `10000` - Simple digits
- `10.000` - Thousands separator (dot)
- `10.000,50` - With decimal separator (comma)
- `1.000.000,99` - Multiple thousands separators

## Invalid Formats (Automatically Prevented)

- `10...1` - Multiple consecutive dots
- `10,1,1` - Multiple commas
- `10,,1` - Consecutive commas
- `10.000.00` - Decimal should use comma, not dot

## Usage

```tsx
import { PriceInput } from '@/forms/price-input'

const [price, setPrice] = useState('')

<PriceInput
  value={price}
  onChange={setPrice}
  label='Amount'
  helperText='Enter price in IDR format (e.g., 10.000 or 10.000,50)'
  error={error}
  allowNegative={false}
/>
```

## Helper Functions

The component includes utility functions for working with Rupiah values:

```tsx
import {
  parseRupiahToNumber,
  formatNumberToRupiah,
  isValidRupiahFormat,
} from '@/forms/price-input/helpers'

// Convert Rupiah string to number
const numericValue = parseRupiahToNumber('10.000,50') // 10000.5

// Format number to Rupiah display
const displayValue = formatNumberToRupiah(10000.5) // '10.000,50'

// Validate Rupiah format
const isValid = isValidRupiahFormat('10.000,50') // true
```
