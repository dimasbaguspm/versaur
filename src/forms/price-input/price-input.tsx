import { useRef, useState } from 'react'
import type { ChangeEvent, FC } from 'react'
import { TextInput } from '@/forms/text-input'
import type { PriceInputProps } from './types'
import { Icon } from '@/primitive/icon'
import { usePriceInputSync } from './use-price-input'
import { sanitizeRupiahInput, isValidRupiahFormat } from './helpers'
import { Banknote } from 'lucide-react'

/**
 * PriceInput is a specialized input component for Indonesian Rupiah (IDR) currency.
 * It enforces proper Rupiah formatting with thousands separators (dots) and decimal separators (commas).
 * Valid formats: 10000, 10.000, 10.000,50, 1.000.000,99
 * Invalid formats are automatically prevented: 10...1, 10,1,1, 10,,1, 10.000.00
 */
export const PriceInput: FC<PriceInputProps> = ({
  value,
  onChange,
  allowNegative = false,
  inputProps,
  label,
  helperText,
  error,
  variant,
  ...rest
}) => {
  const inputRef = useRef<HTMLInputElement>(null)
  const [rawValue, setRawValue] = useState<string>(
    value === '' ? '' : String(value)
  )
  usePriceInputSync(value, setRawValue)

  // Handle input change: validate and sanitize Rupiah format
  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value

    // First sanitize the input to remove obviously invalid characters
    const sanitized = sanitizeRupiahInput(inputValue, allowNegative)

    // Then validate if the sanitized input follows Rupiah format
    if (isValidRupiahFormat(sanitized, allowNegative)) {
      setRawValue(sanitized)
      onChange(sanitized)
    } else {
      // If invalid, keep the previous valid value
      // This prevents invalid states from being entered
      e.preventDefault()
    }
  }

  return (
    <TextInput
      ref={inputRef}
      type='text'
      label={label}
      value={rawValue}
      onChange={handleInputChange}
      error={error}
      helperText={helperText}
      placeholder={rest.placeholder}
      leftContent={<Icon as={Banknote} color='inherit' size='sm' />}
      variant={variant}
      inputMode='numeric'
      autoComplete='off'
      pattern='[0-9.,-]*'
      {...inputProps}
      {...rest}
    />
  )
}
