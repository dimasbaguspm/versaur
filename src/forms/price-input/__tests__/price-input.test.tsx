import { render, screen, fireEvent } from '@testing-library/react'
import { vi } from 'vitest'
import { composeStories } from '@storybook/react'
import * as stories from '../price-input.stories'
import {
  isValidRupiahFormat,
  sanitizeRupiahInput,
  parseRupiahToNumber,
  formatNumberToRupiah,
} from '../helpers'

describe('PriceInput', () => {
  const { Default, WithError } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  describe('Rupiah format validation', () => {
    it('accepts valid Rupiah formats', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      // Valid formats
      const validInputs = ['10000', '10.000', '10.000,50', '1.000.000,99']
      validInputs.forEach(value => {
        fireEvent.change(input, { target: { value } })
        expect(input).toHaveValue(value)
        expect(handleChange).toHaveBeenCalledWith(value)
      })
    })
  })

  it('shows error state', () => {
    render(<WithError />)
    expect(screen.getByText(/invalid price/i)).toBeInTheDocument()
  })

  describe('onChange behavior with various Rupiah formats', () => {
    it('passes correct values for basic number formats', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      const testCases = [
        { input: '1', expected: '1' },
        { input: '10', expected: '10' },
        { input: '100', expected: '100' },
        { input: '1000', expected: '1000' },
        { input: '10000', expected: '10000' },
        { input: '100000', expected: '100000' },
        { input: '1000000', expected: '1000000' },
      ]

      testCases.forEach(({ input: inputValue, expected }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })

    it('passes correct values for thousands separator formats', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      const testCases = [
        { input: '10.000', expected: '10.000' },
        { input: '100.000', expected: '100.000' },
        { input: '1.000.000', expected: '1.000.000' },
        { input: '10.000.000', expected: '10.000.000' },
      ]

      testCases.forEach(({ input: inputValue, expected }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })

    it('passes correct values for decimal formats', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      const testCases = [
        { input: '10000,5', expected: '10000,5' },
        { input: '10000,50', expected: '10000,50' },
        { input: '10.000,5', expected: '10.000,5' },
        { input: '10.000,50', expected: '10.000,50' },
        { input: '1.000.000,99', expected: '1.000.000,99' },
        { input: '0,59', expected: '0,59' },
      ]

      testCases.forEach(({ input: inputValue, expected }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })

    it('passes correct values for intermediate typing states', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      // Test typing "1.000.000,59" step by step
      const typingSequence = [
        { input: '1', expected: '1' },
        { input: '1.', expected: '1.' },
        { input: '1.0', expected: '1.0' },
        { input: '1.00', expected: '1.00' },
        { input: '1.000', expected: '1.000' },
        { input: '1.000.', expected: '1.000.' },
        { input: '1.000.0', expected: '1.000.0' },
        { input: '1.000.00', expected: '1.000.00' },
        { input: '1.000.000', expected: '1.000.000' },
        { input: '1.000.000,', expected: '1.000.000,' },
        { input: '1.000.000,5', expected: '1.000.000,5' },
        { input: '1.000.000,59', expected: '1.000.000,59' },
      ]

      typingSequence.forEach(({ input: inputValue, expected }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })

    it('rejects inputs that start with separators', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      // These should now be rejected since we don't allow starting with separators
      const testCases = [
        { input: ',', expected: null },
        { input: ',5', expected: null },
        { input: ',59', expected: null },
        { input: '.59', expected: null },
      ]

      // First set a valid value
      fireEvent.change(input, { target: { value: '10000' } })
      expect(input).toHaveValue('10000')

      testCases.forEach(({ input: inputValue }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        // These should be rejected and not call onChange
        expect(handleChange).not.toHaveBeenCalled()
        expect(input).toHaveValue('10000') // Should keep previous valid value
      })
    })

    it('handles input sanitization and passes sanitized values', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      const testCases = [
        { input: '10abc000', expected: '10000' },
        { input: '10.000$', expected: '10.000' },
        { input: 'Rp 10.000,50', expected: '10.000,50' },
        { input: '10...000', expected: '10.000' },
        { input: '10..000', expected: '10.000' },
        { input: '10,50,75', expected: '1050,75' },
      ]

      testCases.forEach(({ input: inputValue, expected }) => {
        handleChange.mockClear()
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })

    it('rejects invalid formats and does not call onChange', () => {
      const handleChange = vi.fn()
      render(<Default onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      // Set initial valid value
      fireEvent.change(input, { target: { value: '10000' } })
      expect(input).toHaveValue('10000')
      handleChange.mockClear()

      // These should be rejected after sanitization and validation
      const invalidInputs = [
        '.000', // starts with dot
        ',59', // starts with comma
        '.59', // starts with dot
        ',', // just comma
        '.', // just dot
        '10.0000', // more than 3 digits in thousands segment
      ]

      invalidInputs.forEach(inputValue => {
        fireEvent.change(input, { target: { value: inputValue } })
        expect(input).toHaveValue('10000') // Should keep previous valid value
        expect(handleChange).not.toHaveBeenCalled()
      })
    })

    it('rejects negative values when not allowed', () => {
      const handleChange = vi.fn()
      render(<Default allowNegative={false} onChange={handleChange} />)
      const input = screen.getByLabelText(/amount/i)

      // Set initial valid value
      fireEvent.change(input, { target: { value: '10000' } })
      expect(input).toHaveValue('10000')
      handleChange.mockClear()

      // Negative values should be sanitized to positive
      const testCases = [
        { input: '-10000', expected: '10000' },
        { input: '-10.000,50', expected: '10.000,50' },
      ]

      testCases.forEach(({ input: inputValue, expected }) => {
        fireEvent.change(input, { target: { value: inputValue } })
        expect(handleChange).toHaveBeenCalledWith(expected)
        expect(input).toHaveValue(expected)
      })
    })
  })
})

describe('Rupiah helpers', () => {
  describe('isValidRupiahFormat', () => {
    it('validates correct Rupiah formats', () => {
      const validFormats = [
        '',
        '0',
        '10',
        '100',
        '1000',
        '10000',
        '10.000',
        '100.000',
        '1.000.000',
        '10.000,5',
        '10.000,50',
        '1.000.000,99',
        '0,59',
        // Intermediate typing states
        '10.',
        '1.000.',
        '1.000.000,',
        '10.0',
        '1.000.0',
      ]

      validFormats.forEach(format => {
        expect(isValidRupiahFormat(format)).toBe(true)
      })
    })

    it('rejects invalid Rupiah formats', () => {
      const invalidFormats = [
        '.000', // starts with dot
        ',59', // starts with comma
        '.59', // starts with dot
        ',', // just comma
        '.', // just dot
        '10..000', // consecutive dots
        '10...1', // multiple consecutive dots
        '10,1,1', // multiple commas
        '10,,1', // consecutive commas
        // '10.000.00', // decimal should use comma, not dot
        '10.000,123', // more than 2 decimal digits
        '10.0000', // more than 3 digits in thousands segment
      ]

      invalidFormats.forEach(format => {
        expect(isValidRupiahFormat(format)).toBe(false)
      })
    })

    it('handles negative values when allowed', () => {
      expect(isValidRupiahFormat('-', true)).toBe(true)
      expect(isValidRupiahFormat('-10000', true)).toBe(true)
      expect(isValidRupiahFormat('-10.000,50', true)).toBe(true)

      expect(isValidRupiahFormat('-', false)).toBe(false)
      expect(isValidRupiahFormat('-10000', false)).toBe(false)
    })
  })

  describe('sanitizeRupiahInput', () => {
    it('removes invalid characters', () => {
      expect(sanitizeRupiahInput('10abc000')).toBe('10000')
      expect(sanitizeRupiahInput('10.000$')).toBe('10.000')
      expect(sanitizeRupiahInput('Rp 10.000,50')).toBe('10.000,50')
    })

    it('handles multiple commas correctly', () => {
      expect(sanitizeRupiahInput('10,50,75')).toBe('1050,75')
      expect(sanitizeRupiahInput('10,1,1')).toBe('101,1')
    })

    it('removes consecutive dots', () => {
      expect(sanitizeRupiahInput('10...000')).toBe('10.000')
      expect(sanitizeRupiahInput('10..000')).toBe('10.000')
    })
  })

  describe('parseRupiahToNumber', () => {
    it('converts Rupiah format to number', () => {
      expect(parseRupiahToNumber('10000')).toBe(10000)
      expect(parseRupiahToNumber('10.000')).toBe(10000)
      expect(parseRupiahToNumber('10.000,50')).toBe(10000.5)
      expect(parseRupiahToNumber('1.000.000,99')).toBe(1000000.99)
    })

    it('handles negative values', () => {
      expect(parseRupiahToNumber('-10.000,50')).toBe(-10000.5)
    })

    it('handles empty and invalid inputs', () => {
      expect(parseRupiahToNumber('')).toBe(0)
      expect(parseRupiahToNumber('-')).toBe(0)
    })
  })

  describe('formatNumberToRupiah', () => {
    it('formats number to Rupiah display format', () => {
      expect(formatNumberToRupiah(10000)).toBe('10.000')
      expect(formatNumberToRupiah(10000.5)).toBe('10.000,5')
      expect(formatNumberToRupiah(1000000.99)).toBe('1.000.000,99')
    })

    it('handles negative values', () => {
      expect(formatNumberToRupiah(-10000.5)).toBe('-10.000,5')
    })

    it('handles zero', () => {
      expect(formatNumberToRupiah(0)).toBe('')
    })
  })
})
