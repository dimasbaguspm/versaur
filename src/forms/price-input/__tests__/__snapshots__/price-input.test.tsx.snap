// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PriceInput > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-foreground mb-2"
      for="«r0»"
    >
      Amount
    </label>
    <div
      class="relative w-full"
    >
      <span
        aria-hidden="true"
        class="absolute left-2.5 top-0 bottom-0 pointer-events-none text-gray-500 flex items-center justify-center w-5"
        data-testid="left-content"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-banknote inline-flex items-center justify-center w-4 h-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="12"
            rx="2"
            width="20"
            x="2"
            y="6"
          />
          <circle
            cx="12"
            cy="12"
            r="2"
          />
          <path
            d="M6 12h.01M18 12h.01"
          />
        </svg>
      </span>
      <input
        aria-describedby="«r0»-helper"
        aria-invalid="false"
        autocomplete="off"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:pointer-events-none disabled:bg-gray-50 border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 pl-9 pr-3 h-9"
        id="«r0»"
        inputmode="numeric"
        pattern="[0-9.,-]*"
        type="text"
        value=""
      />
    </div>
    <div
      class="mt-1 text-sm text-gray-600"
      id="«r0»-helper"
    >
      Enter price in IDR format (e.g., 10.000 or 10.000,50)
    </div>
  </div>
</DocumentFragment>
`;
